app/browser/api/ledger.js

```diff
@@ -0,0 +1,2187 @@
+    if (err) {
+      console.error(err)
+    } else {
+      tabs.create({url: fileUrl(filePath)}, (webContents) => {
+        if (backupAction === 'print') {
+          webContents.print({silent: false, printBackground: false})
+        } else {
+          webContents.downloadURL(fileUrl(filePath), true)
+        }
+      })
+    }
+  })
+}
+
+const recoverKeys = (state, useRecoveryKeyFile, firstKey, secondKey) => {
+  let firstRecoveryKey, secondRecoveryKey
+
+  if (useRecoveryKeyFile) {
+    let recoveryKeyFile = promptForRecoveryKeyFile()
+    if (!recoveryKeyFile) {
+      // user canceled from dialog, we abort without error
+      return state
+    }
+
+    if (recoveryKeyFile) {
+      const result = loadKeysFromBackupFile(state, recoveryKeyFile)
+      const keys = result.keys || {}
+      state = result.state
+
+      if (keys) {
+        firstRecoveryKey = keys.paymentId
+        secondRecoveryKey = keys.passphrase
+      }
+    }
+  }
+
+  if (!firstRecoveryKey || !secondRecoveryKey) {
+    firstRecoveryKey = firstKey
+    secondRecoveryKey = secondKey
+  }
+
+  const UUID_REGEX = /^[0-9a-z]{8}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{4}-[0-9a-z]{12}$/
+  if (
+    typeof firstRecoveryKey !== 'string' ||
+    !firstRecoveryKey.match(UUID_REGEX) ||
+    typeof secondRecoveryKey !== 'string' ||
+    !secondRecoveryKey.match(UUID_REGEX)
+  ) {
+    // calling logError sets the error object
+    state = logError(state, true, 'recoverKeys')
+    state = ledgerState.setRecoveryStatus(state, false)
+    return state
+  }
+
+  client.recoverWallet(firstRecoveryKey, secondRecoveryKey, (err, result) => {
+    appActions.onWalletRecovery(err, result)
+  })
+
+  return state
+}
+
+const onWalletRecovery = (state, error, result) => {
+  let existingLedgerError = ledgerState.getInfoProp(state, 'error')
+
+  if (error) {
+    // we reset ledgerInfo.error to what it was before (likely null)
+    // if ledgerInfo.error is not null, the wallet info will not display in UI
+    // logError sets ledgerInfo.error, so we must we clear it or UI will show an error
+    state = logError(error, 'recoveryWallet')
+    state = ledgerState.setInfoProp(state, 'error', existingLedgerError)
+    state = ledgerState.setRecoveryStatus(state, false)
+  } else {
+    callback(error, result)
+
+    if (balanceTimeoutId) {
+      clearTimeout(balanceTimeoutId)
+    }
+    getBalance(state)
+    state = ledgerState.setRecoveryStatus(state, true)
+  }
+
+  return state
+}
+
+const quit = (state) => {
+  quitP = true
+  state = addVisit(state, locationDefault, new Date().getTime(), null)
+
+  if (!getSetting(settings.PAYMENTS_ENABLED) && getSetting(settings.SHUTDOWN_CLEAR_HISTORY)) {
+    state = ledgerState.resetSynopsis(state, true)
+  }
+
+  return state
+}
+
+const initSynopsis = (state) => {
+  state = ledgerState.saveSynopsis(state, null, synopsis.options)
+  let value = getSetting(settings.PAYMENTS_MINIMUM_VISIT_TIME)
+  if (!value) {
+    value = 8 * 1000
```

Comment:
maybe we can pull this magic number up to the top 😄

Reply:
done